### PR TITLE
Feature : Item Name Scrolling

### LINKED ISSUE #85: OLED Scrolling
**Is your feature request related to a problem? Please describe.**
Psudo problem: I used discord frequently and have ducking enabled. This changes the active volume of all my audio sessions when someone is speaking. This in turn triggers the display to wake up, even if the display is currently on the Master volume which does not change. So the result is the device is actively waking up very frequently and not much I can do about it. This is also not very great for the OLED screens as they are extremely sensitive to burn in.

**Describe the solution you'd like**
The display should have the ability to scroll content while displaying to help reduce burn-in. Plus a good scrolling effect always looks great. The SSD1306 has hardware scrolling ability, however the display cannot be updated while this is scrolling. So we would need to implement a buffer/virtual scrolling option or go with an approach similar to what was done for the QMK Keyboards: Use hardware scrolling until the contents of the screen change, then turn off hardware scrolling to update the screen, and wait for another idle screen timeout to begin scrolling again.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by SchemingWeasels**
## Issues
 - Closes #85 

## Description
Long application names now scroll on the screen after a delay.

## Types of changes
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [X] Requested changes are in a branch
- [X] Compiled and tested requested changes on target hardware (PC, device)
- [ ] Updated the documentation, if necessary
- [ ] Updated the LICENSES file, if necessary
- [X] Reviewed the [guidelines for contributing](../CONTRIBUTING.md) to this repository

## Further comments

Commits include changes from #125 as it hasn't been merged into master, and is a required dependency for this change.

### CODE REVIEW

**Comment by SchemingWeasels** [comment]
There a bug with game edit mode. If you wake the screen, sometimes the scrolling just messes everything up. Can't seem to figure out why.

**Comment by t3knomanzer** [comment]
I haven't looked at the implementation, I'll leave that to @XScorpion2, but I pulled the code and ran it. 
Man, it's so cool to see stuff moving on the screen! I did up the steps to 30 to slow it down a bit, I like it a lot better like that.

**Review by XScorpion2 — APPROVED** [review]
First approach looks good. There is a bit of work / rework on how the display code is handled to really make it scale as currently the full frame is being updated every loop, which is not necessary as the oled itself has a buffer that we only need to push changes to instead of the full frame.

**Comment by SchemingWeasels** [comment]
> There a bug with game edit mode. If you wake the screen, sometimes the scrolling just messes everything up. Can't seem to figure out why.

I would hold on merging till this is figured out. Although I have some work that needs some rather urgent attention for the next few days so I apologize if I'm blocking/holding anyone or anything up.

**Comment by t3knomanzer** [comment]
@XScorpion2 I don't see any method in the Adafruit SSD1306 library to do partial updates of the screen.
We can certainly send commands and data so it'd be a matter of just creating our own partial update function and calling that one instead of display->display();

Ref: 
https://electronics.stackexchange.com/questions/335152/ssd1306-i2c-draw-only-changes-works-however-on-wrong-position-how-to-set-posit

https://github.com/adafruit/Adafruit_SSD1306

**Comment by XScorpion2** [comment]
@t3knomanzer ya, would need to add that function. 
And as it just so has it, I did just that for the qmk keyboard firmware: 
https://github.com/qmk/qmk_firmware/blob/master/drivers/oled/oled_driver.c 
I should be able to take that and modify it for use with Arduino's wire library.

**Comment by t3knomanzer** [comment]
@XScorpion2 That's fantastic! Actually, on another note... we should also start taking all those other .ino files and making them into real C libraries.  
I'm not super experienced in C/C++ so there is a lot of room for improvement in many areas. From code organization to style and memory management.  
I'm definitely going to take a look at the qmk_firmware and use that as reference. I'm glad the oled driver is a pure C library and not C++. IMO, C++ is overkill for MCUs but that's just an opinion.

**Comment by t3knomanzer** [comment]
@XScorpion2 Do you have any idea when you'll be able to tackle this? It's the only task left to release 1.1.0. No pressure of course, we all have day jobs :D

**Comment by XScorpion2** [comment]
I've set aside some time this weekend to work on it. Not sure if the right approach is to modify the qmk oled code, or to modify the adafruit library at this point, but either way it should be about the same amount of time to knock out.

**Comment by t3knomanzer** [comment]
Awesome!
I really like the code in QMK.   
I personally would love to go that way. Get rid of all the third party libs we’re using and and make our own C implementations but that’s just personal preference.   
We can do one at a time of course. 

By the looks of it, the oled driver in qmk is a full implementation, completely standalone, right?

File: Desktop/Application/MaxMix/MaxMix.csproj
Comment (SchemingWeasels) [inline Desktop/Application/MaxMix/MaxMix.csproj]: Ugh this is still haunting me haha

**Comment by SchemingWeasels** [comment]
Just to confirm, are you guys getting the crashes while in game edit mode with this PR? I've noticed that it happens a lot more with "Steam Client Bootstrapper"

**Comment by t3knomanzer** [comment]
@SchemingWeasels I'm seeing this.

**Comment by SchemingWeasels** [comment]
Thanks. Pulling my hair out at the moment. The drawing code is identical to the larger font size (Nav, App, Game Select Screens) but this crashes while the other doesn't.

I doubt it's font size related. App edit is also font size 1 and doesn't crash.

Perhaps it's a ram issue... my code is hella sloppy and game edit mode draws tons of stuff. 

Edit: DOH~! I didn't pass by reference in DisplayGameEditScreen. That should do it...

**Comment by t3knomanzer** [comment]
I tried reducing the max number of items to see if we were running out of memory but doesn't seem to be the case.

I'm working on something else now but I can help you debug it over the weekend maybe.

**Comment by SchemingWeasels** [comment]
> I tried reducing the max number of items to see if we were running out of memory but doesn't seem to be the case.
> 
> I'm working on something else now but I can help you debug it over the weekend maybe.

No worries. I figured it out. You can see the progression of catching the bug in my previous edited comment 🤦.

1. What values do you use for the step interval delay and do you have a suggestion for a default value? 

2. Should this be editable in the app? I think it is unnecessary. 

3. Do you think we should raise the default sleep time because the display hardly has time to scroll before it goes to sleep with the current default sleep time.

**Comment by t3knomanzer** [comment]
> I did up the steps to 30 to slow it down a bit, I like it a lot better like that.

I can't remember what I increased the sleep to. You choice :D

**Comment by SchemingWeasels** [comment]
I'll push the fix tomorrow. I still have to work a bit on the scrolling in game mode because we are going to scroll 2 items. Haven't decided on the expected behavior yet.

**Comment by t3knomanzer** [comment]
Sounds good! Thanks for the good work @SchemingWeasels

**Comment by t3knomanzer** [comment]
Just to be clear, we are waiting on 2 things here, the fix from @SchemingWeasels and the display code from @XScorpion2.  
Or should we just release it with the fix from @SchemingWeasels and then release the optimizations later?  

I think it all depends on 2 things, how bad is the performance with the current display update code and what does @XScorpion2 have in mind.  
Obviously if it's unsusable performance wise we can't release it but that's not what I have observed in the little testing I've done.  
As far as the update code, I'd love to hear opinions on completely switching the Adafruit libraries for the one @XScorpion2 wrote.  
My vote is on switching it completely, I really like the approach of going pure C. I'd love to slowly get rid of all third party libraries we are using and switching them for C implementations.  
Then once we have switched the libraries we can start making the multiple .ino files we have into libraries too.  
Ultimately completely getting rid of the arduino framework and going full avr-libc.  
The benefits in my opinion are smaller size, cleaner code, less dependencies, less memory usage, better performance, good learning opportunity for some of us.

I'm all ears!

File: Embedded/MaxMix/Display.ino
Comment (SchemingWeasels) [inline Embedded/MaxMix/Display.ino:124]: The +1 here is somewhat a "magic number" but I'm not sure if it makes sense to define a const just for this. 
The +1 here is caused by the side margin (DISPLAY_AREA_CENTER_MARGINE_SIDE = 11) max char width (DISPLAY_CHAR_MAX_WIDTH_X1 = 107) not adding up to the display width (128). 

The way I thought it should work would be DISPLAY_WIDTH = 2 (MARGIN_SIDE) + CHAR_MAX_WIDTH
But currently, if you calculate this you'll get 128 = 129, and thus the +1 to offset the mask.

Thoughts?

**Review by SchemingWeasels — COMMENTED** [review]
Fixed the bug and redid the way scrolling works. I think it's a whole lot better now, what do you think?
Using 'now' from loop instead of calling millis() when handling scrolling.
Changed the default scroll speed to 20/30.
Didn't change the default sleep time. I think it should be a separate PR if we really want to change it.

File: Embedded/MaxMix/Display.ino
Comment (XScorpion2) [inline Embedded/MaxMix/Display.ino]: calculate the new scrollOffset before clearing the display. If the scroll offset has not changed from last call, return. If it has, store the results and continue.

**Comment by XScorpion2** [comment]
So the main performance hit by this change is going to be flushing the display every loop. Adding a check based on the calculated scroll offset before doing any display code, and exiting if it is the same as last call will restore quite a bit of performance. From there I would say if the performance is good (IE: using it feels fine), then to merge it and I'll continue with my performance changes in Display.ino, and possible make modifications to the Adafruit library as well. Based on the usage of the library, I think my oled driver code will be a poor fit as it was designed with different restrictions and requirements in mind.

File: Embedded/MaxMix/Display.ino
Comment (XScorpion2) [reply Embedded/MaxMix/Display.ino]: Note: This should be done in the other locations as well where you are scrolling.

**Comment by t3knomanzer** [comment]
Alright then. I’ll test it in a few hours and merge if it’s fine. Can you guys test it on your side too?

**Comment by SchemingWeasels** [comment]
The display only redraws when it has to now, this change drastically reduces the number of display updates.  
Didn't notice a performance issue before, so I couldn't feel any difference with this change either. It should work exactly like before unless I missed something.

**Comment by t3knomanzer** [comment]
@SchemingWeasels I'm not sure about the reverse scroll. I don't find it useful reading backwards really. Personally I prefer the original implementation where it just reset back after done scrolling.

Thoughts?

Also, there is a bug in Game mode, the acceleration doesn't work. It seems to be caused by the reverse direction detection code, since both items move in opposite directions.

I can take care of that, no worries.

```
else if (stateGame == STATE_GAME_EDIT)
    {
      items[itemIndexGameA].volume = ComputeAcceleratedVolume(encoderDelta, deltaTime, items[itemIndexGameA].volume);
      items[itemIndexGameB].volume = ComputeAcceleratedVolume(-encoderDelta, deltaTime, items[itemIndexGameB].volume);

      SendItemVolumeCommand(&items[itemIndexGameA], sendBuffer, encodeBuffer);
      SendItemVolumeCommand(&items[itemIndexGameB], sendBuffer, encodeBuffer);
    }
```

**Comment by t3knomanzer** [comment]
@SchemingWeasels I have the acceleration fixed in another PR. As soon as you finish this one I'll merge it, rebase the fix and merge it too, then I'll trigger the build for the 1.1.0 release.

**Comment by SchemingWeasels** [comment]
@t3knomanzer I found that the way the item name just suddenly appears to be very jarring. I have another idea. I'm thinking of scrolling past the last char and looping back around to the start of the name. Sorta "continuous scroll" for the name.

For example 
[SPOTIFY PREM]
[POTIFY PREMI]
[OTIFY PREMIU]
:
[PREMIUM    SP]
[REMIUM   SPO]
[EMIUM   SPOT]
:
[SPOTIFY PREM]

Then we'll be back to 2 delays only, one for intervals between scrolls and the other is the delay between steps. Not sure how I wanna implement this yet though. Perhaps some coffee would bring ideas. 

Do you think this would be more intuitive?

**Comment by t3knomanzer** [comment]
Absolutely, that’s be the perfect solution. Scroll all the way to the initial position the pause for a bit and scroll again. 

Thanks for the work!

File: Embedded/MaxMix/Display.ino
Comment (SchemingWeasels) [inline Embedded/MaxMix/Display.ino:477]: So the spacing between characters is not constant, it depends on what the characters are, the font size, and font style etc. Calculating manually will cause some weird edge cases to return wrong offset/calculation values. For example, applications with multiple symbols or special characters like tab and space.

To keep things compatible with any app name even weird ones, which I think is a fair assumption that there can and will be weird names, I used **display->getTextBounds** which is part of the GFX lib.

**Review by SchemingWeasels — COMMENTED** [review]
Started running into SRAM issues. I put all the consts into flash and that helped a bit. I ran into SRAM issues when I joined the names back to back into a long string.

In my opinion, shouldn't most of these be precompiler directives (#define) ? 

Also, I made the separator customizable. It's currently 3 spaces, but I tried with some other chars and it looked alright as well.

Honestly, I'm not really stoked on the way I did it. Feels really sloppy, evident by  the running out of SRAM.

**Comment by t3knomanzer** [comment]
Why don’t you just draw the same string twice instead of modifying the string itself?

Ruben Henares

**Comment by SchemingWeasels** [comment]
@t3knomanzer I do. I modify the string to parse into getTextBounds. Drawing is just using the same string twice if the name is scrollable.

**Comment by t3knomanzer** [comment]
Sorry I haven’t been able to look at this. I’ve had some unexpected things to take care of today.

**Comment by t3knomanzer** [comment]
Hey,

First of all, sorry for the delay, I've had some unexpected things to deal with.
After reviewing this I think that we've iterated on this so much that's become a little messy so I took the ideas from this implementation and re-implemented it from scratch in PR #141.

Thank you all for the awesome work!

**Pull request closed without merging**
